### PR TITLE
Issue #82 - Making validation errors more type-specific & adding some unit tests

### PR DESCRIPTION
**Pull request by robons**

### CODE REVIEW

File: csvqb/csvqb/models/validationerror.py [hunk 1]
```diff
@@ -2,11 +2,19 @@
...
-        self.message: str = message
+    """Class representing an error validating a model."""
 
-    def __repr__(self) -> str:
-        return f"{self.__class__.__name__}({self.message})"
+    message: str
+
+
+@dataclass
+class SpecificValidationError(ValidationError, ABC):
```
Comment (canwaf) [inline csvqb/csvqb/models/validationerror.py:17]: Speaking of nonsense?

Comment (robons) [reply csvqb/csvqb/models/validationerror.py:17]: What?

File: csvqb/csvqb/tests/unit/cube/qb/test_cubeqb_errorvalidation.py [hunk 2]
```diff
@@ -111,10 +120,329 @@ def test_existing_dimension_output_uri_template():
...
+    assert error.component_type == QbMultiUnits
+    assert error.maximum_number == 1
+    assert error.actual_number == 2
+
+
+def test_multiple_obs_val_columns():
+    """
+    Ensure that when a user defines multiple observation value columns, we get an error.
+
+    We only currently accept the `MeasureDimension` style of multi-measure datasets.
```
Comment (canwaf) [inline csvqb/csvqb/tests/unit/cube/qb/test_cubeqb_errorvalidation.py:271]: :soon:

File: csvqb/csvqb/tests/unit/cube/qb/test_cubeqb_errorvalidation.py [hunk 3]
```diff
@@ -111,10 +120,329 @@ def test_existing_dimension_output_uri_template():
...
+    assert isinstance(error, MaxNumComponentsExceededError)
+    assert error.component_type == QbMultiMeasureDimension
+    assert error.maximum_number == 1
+    assert error.actual_number == 2
+
+
+def test_measure_dimension_with_single_measure_obs_val():
+    """
+    Ensure that when a user defines a measure dimension with a single-measure observation value, they get an error.
+    """
```
Comment (canwaf) [inline csvqb/csvqb/tests/unit/cube/qb/test_cubeqb_errorvalidation.py:405]: Why is this an error? It's verbose but is it wrong?

Comment (robons) [reply csvqb/csvqb/tests/unit/cube/qb/test_cubeqb_errorvalidation.py:405]: I see where you're coming from, however, I think they've fundamentally misunderstood something if they're taking this approach so I want to push back on them doing this. 

Don't forget that all of these validation errors are strictly advisory and don't stop the user actually doing anything.

**Review by canwaf — APPROVED** [review]
I look forward to trying to use dataclasses and pylance in my rewrite of #80 now that we've refactored QbAttribute and added a bunch of new errors. It would be good to see a consolidated list of errors so that they can be easily summoned when writing unit tests. I haven't checked if this shows up in a useful way in our docs.

**Comment by robons** [comment]
> I haven't checked if this shows up in a useful way in our docs.

I forgot to include titles in the files so it isn't that clear where they are... but there's in there. They appear as if from nowhere. I'll fix and recommit.
image

**Pull request merged**